database/002_initial_schema.sql

```diff
@@ -1,5 +1,470 @@
+  CHECK (value ~ '^[a-zA-Z0-9.!#$%&''*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$' );
+
+-- ENUMS
+
+CREATE TYPE statues AS ENUM ('0', '1', '2');
+CREATE TYPE Etypes AS ENUM('itex','booking','ad');
+-- CONFIG
+REVOKE CONNECT ON DATABASE booking_dev FROM PUBLIC;
+REVOKE ALL ON ALL TABLES IN SCHEMA public FROM PUBLIC;
+
+-- Tables
+
+-- Company
+
+CREATE TABLE
+IF NOT EXISTS public.company(
+    CID SERIAL PRIMARY KEY,
+    email citext UNIQUE NOT NULL,
+    orgNum bigint UNIQUE NOT NULL,
+    username text NOT NULL ,
+    salt text NOT NULL ,
+    passhash varchar NOT NULL ,
+    fullName text UNIQUE NOT NULL
+);
+
+-- Company User
+CREATE TABLE
+IF NOT EXISTS public.companyuser(
+    CUID SERIAL PRIMARY KEY,
+    email citext UNIQUE,
+    telephone text,
+    username text NOT NULL,
+    salt text NOT NULL,
+    passhash varchar NOT NULL ,
+    fullName text UNIQUE NOT NULL,
+    CID INTEGER REFERENCES company(CID) NOT NULL
+);
+
+--OWUser
+CREATE TABLE
+IF NOT EXISTS public.onlineuser(
+    OID SERIAL PRIMARY KEY,
+    JWT varchar NOT NULL,
+    companies int[]
+);
+
+--Event
 CREATE TABLE
-IF NOT EXISTS public.keys
-(accesstoken text UNIQUE);
+IF NOT EXISTS public.event(
+    EID SERIAL PRIMARY KEY,
+    statusCode statues NOT NULL
+    );
+
+--Offer
+CREATE TABLE
+IF NOT EXISTS public.offer(
+    OFID SERIAL PRIMARY KEY,
+    statusCode statues NOT NULL,
+    data jsonb NOT NULL,
+    EID INTEGER REFERENCES event(EID) NOT NULL,
+    CID INTEGER REFERENCES company(CID) NOT NULL,
+    OID INTEGER REFERENCES onlineuser(OID) NOT NULL
+);
+
+--Wish
+CREATE TABLE
+IF NOT EXISTS public.wish(
+    WID SERIAL PRIMARY KEY,
+    statusCode statues NOT NULL,
+    data JSONB NOT NULL,
+    EID INTEGER REFERENCES event(EID) NOT NULL,
+    CUID INTEGER REFERENCES companyuser(CUID),
+    CID INTEGER REFERENCES company(CID) NOT NULL
+);
+--Contracts
+CREATE TABLE
+IF NOT EXISTS public.contracts(
+    contractID SERIAL PRIMARY KEY,
+    acceptDate bigint NOT NULL,
+    expirationDate bigint NOT NULL,
+    price bigint NOT NULL,
+    billingAdresse text NOT NULL,
+    statusCode statues NOT NULL,
+    path text NOT NULL,
+    orgNum bigint REFERENCES company(orgNum) NOT NULL,
+    EID INTEGER REFERENCES event(EID) NOT NULL,
+    CUID INTEGER REFERENCES companyuser(CUID),
+    CID INTEGER REFERENCES company(CID) NOT NULL,
+    OID INTEGER REFERENCES onlineuser(OID) NOT NULL
+);
+--Ads
+CREATE TABLE
+IF NOT EXISTS ad(
+    CID INTEGER REFERENCES company(CID) NOT NULL
+    ) INHERITS (event)
+;
+-- Queue
+CREATE TABLE
+IF NOT EXISTS public.queue(
+    QID SERIAL PRIMARY KEY,
+    statusCode INTEGER NOT NULL,
+    --Epoch Time Format
+    enteredQueue bigint
+        NOT NULL DEFAULT
+        (date_part
+        ('epoch'::text, now
+        ()) *
+        (1000)::double precision),
+    CID INTEGER REFERENCES company(CID) NOT NULL,
+    EID INTEGER REFERENCES event(EID) NOT NULL
```

Comment:
This tab is a queue for companies that has requested an event with Online. The reason for the statusCode is to check if they requested, got an offer, accepted, gotten rejected or rejected our offer. 

I have no other cases for the queue atm, but it could be could be expanded to allow for queueing those who have requested ads in Offline.

I can check if that is easily to implement in postgreSQL and comment on this thread.

Reply:
Gotcha, 

I think a form of timestamp of just an incremental counter for each new table row will be enough for a queue. 

The queue is basically there to sort the companies/offers/requests, right? So a counter will have the benefit of us easily being able to re-arrange the queue/sorted list. in case we want to prioritize a company not only based on when they sent the request. Usually during the booking process, the order of requests being sent is not the most important thing. 

If there is some additional functionality in this queue-tab it can probably be useful. If not, my initial thought is that an extra attribute used to sort them is enough. The other attributes, such as status, will allow us to filter out those that are not relevant. i.e. the requests that have received offers already etc.

Reply:
Counters in postgres does not seem to be easy, nor clean programatically. I think we should explore other options. 

I would rather have a priority flag on each row, and sort first by timestamp, and then by priority to get the correct ordering for companies. 

Is that ok with you?